pkg/apicheck/check.go

```diff
@@ -167,15 +180,43 @@ func (c *ApiConnectivityCheck) handleError() bool {
 			if apiErrorsResponsesSum > nrAllNodes/2 { //already reached more than 50% of the nodes and all of them returned api error
 				//assuming this is a control plane failure as others can't access api-server as well
 				c.config.Log.Info("More than 50% of the nodes couldn't access the api-server, assuming this is a control plane failure")
-				return true
+				return peers.Response{IsHealthy: true, Reason: peers.HealthyBecauseMostPeersCantAccessAPIServer}
 			}
 		}
 
 	}
 
 	//we asked all peers
 	c.config.Log.Error(fmt.Errorf("failed health check"), "Failed to get health status peers. Assuming unhealthy")
-	return false
+	return peers.Response{IsHealthy: false, Reason: peers.UnHealthyBecauseNodeIsIsolated}
+}
+
+func (c *ApiConnectivityCheck) canOtherMastersBeReached() bool {
+	nodesToAsk := c.config.Peers.GetPeersAddresses(peers.Master)
+	numOfMasterPeers := len(nodesToAsk)
+	if nodesToAsk == nil || numOfMasterPeers == 0 {
+		c.config.Log.Info("Peers list is empty and / or couldn't be retrieved from server, other masters can't be reached")
+		return false
+	}
+	chosenNodesAddresses := c.popNodes(&nodesToAsk, numOfMasterPeers)
```

Comment:
ping @mshitrit

Reply:
>why do you introduce a WaitGroup?  

You are correct it is not needed, I've added it when I've done some debugging and it remained.
However I don't think that the comparison to `sumPeersResponses` is correct since `sumPeersResponses` is essintially a synced method and `canOtherControlPlanesBeReached` isn't.
The reason that `canOtherControlPlanesBeReached` would work without the WaitGroup because it would be blocked by the channel access that happens in `sumPeersResponses` .
In any case I think the blocking is much clearer with the WaitGroup so I'd prefer to keep it.

> why tmpAddress? (I'm aware of the potential issues with for loops, but I don't think it's relevant for a string slice and passing the string by value to the next function)

It is indeed relevant in this case.
You verify it running [this](https://go.dev/play/p/HD5q1sDtnHo) snippet in Go Playground.
However I can't take credit for spotting it the linter was smart enough to mark this as an issue :smile: 

>But what I'm wondering most about is, why fixing potential issues here and leaving them at the other place?

I might have missed it, where do you mean ?

Reply:
> However I don't think that the comparison to sumPeersResponses is correct since sumPeersResponses is essintially a synced method and canOtherControlPlanesBeReached isn't.

I didn't want to compare those, but show that `sumPeersResponses` is waiting until all requests returned. I'm comparing part of `getWorkerPeersResponse` with `canOtherControlPlanesBeReached`

> In any case I think the blocking is much clearer with the WaitGroup so I'd prefer to keep it.

maybe, but then we are at my comment:

> But what I'm wondering most about is, why fixing potential issues here and leaving them at the other place?

You are copying code from `getWorkerPeersResponse`, and improving it, maybe even fixing a serious bug (tmpAddress). Why don't you fix the code in  `getWorkerPeersResponse` as well?? Or even better, why don't you extract shared code into a function, so that both functions benefit, and we don't have duplicated code?

Reply:
I do not understand atm why, but `go doSth(address)` vs `go func(){doSth(address)}()` makes the difference in the loop 🤷🏼‍♂️

Reply:
And here is the answer: `go doSth(address)` is similar to the 2nd example here: https://go.dev/doc/faq#closures_and_goroutines
`address` as argument is evaluated for every iteratiom and passed by value to `doSth(address)`.
While with `go func(){doSth(address)}()` `address` is evaluated later when the async func is running.

However, my issue isn't that much that you want to be on the safe side and introduce a new var, it is to fix it at one place only.

Reply:
>Why don't you fix the code in getWorkerPeersResponse as well??

There is no bug in `getWorkerPeersResponse` , the difference is that in `canOtherControlPlanesBeReached` an anonymous go routine is used.

> why don't you extract shared code into a function, so that both functions benefit, and we don't have duplicated code?

IMO both method are very different, `getWorkerPeersResponse` is  71 lines and `canOtherControlPlanesBeReached` is 26 lines,they have a different return val an I can see very little code they have in common, they do share `GetPeersAddresses` which indeed was refactored to support both.

Reply:
> the difference is that in canOtherControlPlanesBeReached an anonymous go routine is used.

I see, but I absolutely do not understand  why??

`getWorkerPeersResponse` has this:
```
		nrAddresses := len(chosenNodesAddresses)
		responsesChan := make(chan selfNodeRemediation.HealthCheckResponseCode, nrAddresses)

		for _, address := range chosenNodesAddresses {
			go c.getHealthStatusFromPeer(address, responsesChan)
		}

		healthyResponses, unhealthyResponses, apiErrorsResponses, _ := c.sumPeersResponses(nrAddresses, responsesChan)
```
`canOtherControlPlanesBeReached` has this:
```
	nrAddresses := len(chosenNodesAddresses)
	responsesChan := make(chan selfNodeRemediation.HealthCheckResponseCode, nrAddresses)
	var wg sync.WaitGroup
	wg.Add(len(chosenNodesAddresses))
	for _, address := range chosenNodesAddresses {
		tmpAddress := address
		go func() {
			defer wg.Done()
			c.getHealthStatusFromPeer(tmpAddress, responsesChan)
		}()

	}
	wg.Wait()
	//We are not expecting API Server connectivity at this stage, however an API Error is an indication to communication with the peer (peer is communicating with current node that it was unable to reach the API server)
	_, _, apiErrorsResponses, _ := c.sumPeersResponses(nrAddresses, responsesChan)
```
That's the same funtionality with slightly different code.
Why the differences?
Why not extracting this into a shared func?

Reply:
>Why not extracting this into a shared func?

No objections, but excluding `WaitGroup` (which is really a NIT) and `c.sumPeersResponses` (that utilize different return vals)  we are talking about 4 lines of common code, I consider that a NIT at best.
I'll update the code :-) 

>I see, but I absolutely do not understand why??

TBH I don't have a good answer, my best guess is that it has to do with the scope created by the anonymous method - feel free to investigate though.

Reply:
>> I see, but I absolutely do not understand why??

> TBH I don't have a good answer, my best guess is that it has to do with the scope created by the anonymous method - feel free to investigate though.

the "why" was about why you do the same thing but in a different way :)

Reply:
even when it's a few lines only, the fact that you do things slightly different than at the other place shows that those lines aren't trivial, so I prefer to have them once only and not twice.